ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,772 @@
+            z_right,
+        }
+    }
+
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.x_left.len()
+    }
+
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.x_left.is_empty()
+    }
+}
+
+/// `SegmentEntry` is a simple wrapper to represent one entry of a `Segment`
+/// currently, we only support `BitSlices`
+#[derive(Clone, Debug)]
+pub struct SegmentEntry<'a>(&'a BitSlice<u8, Lsb0>);
+
+impl<'a> SegmentEntry<'a> {
+    #[must_use]
+    pub fn new(entry: &'a BitSlice<u8, Lsb0>) -> Self {
+        SegmentEntry(entry)
+    }
+
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+/// `UnverifiedValuesStore` stores a vector of `UnverifiedValues` together with an `offset`.
+/// `offset` is the minimum `RecordId` for the current batch.
+/// `chunk_size` is the amount of records within a single batch. It is used to estimate the vector length
+/// during the allocation.
+/// `segment_size` is an option and set to `Some` once we receive the first segment and thus know the size.
+///
+/// `vec` is a `Vec<Option>` such that we can initialize the whole vector with `None` and access any
+/// location out of order. This is important since we do not now the order in which different records
+/// are added to the `UnverifiedValueStore`.
+/// It is not to keep track whether all relevant records have been added. The verification will fail
+/// when a helper party forgets to include one of the records. We could add a check separate from the proof
+/// such that this does not count as malicious behavior.
+#[derive(Clone, Debug)]
+struct UnverifiedValuesStore {
+    offset: RecordId,
+    chunk_size: usize,
+    segment_size: Option<usize>,
+    vec: Vec<Option<UnverifiedValues>>,
+}
+
+#[allow(dead_code)]
+impl UnverifiedValuesStore {
+    /// Creates a new store for given `chunk_size`. `offset` is initialized to `0`.
+    /// Lazy allocation of the vector. It is allocated once the first segment is added.
+    /// `segment_size` is set to `None` since it may not be known yet
+    fn new(chunk_size: usize) -> Self {
+        Self {
+            offset: RecordId::from(0usize),
+            chunk_size,
+            segment_size: None,
+            vec: Vec::<Option<UnverifiedValues>>::new(),
+        }
+    }
+
+    /// `update` updates the current store to the next offset which is the previous offset plus `vec.actual_len()`.
+    /// It deallocates `vec`. `chunk_size` remains the same.
+    ///
+    /// ## Panics
+    /// Panics when `segment_size` is `None`
+    fn update(&mut self) {
+        // compute how many records have been added
+        // since `actual_length` is imprecise (i.e. rounded up) for segments smaller than 256,
+        // subtract `1` to make sure that offset is set sufficiently small
+        self.offset += ((self.actual_len() - 1) << 8) / self.segment_size.unwrap();
```

Comment:
I do not understand this computation. You take the "actual length", subtract 1, then bitshift it up by 8? Why?

Reply:
This has to do with the translation between record_id and index in the vector. A record can have any bit length, from 1 bit to 256+ bits (the latter when vectorized). We translate this into pieces of 256 bit. To compute the new offset, i.e. setting the offset for the next proof chunk, I need to compute the last record_id of the previous chunk which I can compute by computing the len(), multiply it by 256 (shift by 8) and divide by the segment length. Since in case of records with length less than 256, I pack several records in the same index. Since the last index might not be completely filled, I need to be conservative and set the offset lower to ensure all record ids of the following proof chunk are larger than the offset.

Reply:
notice that I use an Vector<Option<>> which is always filled with None unless it has an actual value. Actual length just computes the index of the last some element plus 1.

Reply:
That is a good point. It should only do something when it is not empty. I have changed that.